src/qutip_qip/decompose/decompose_general_qubit_gate.py

```diff
@@ -0,0 +1,466 @@
+    U = input_array
+    check_gate(Qobj(U), num_qubits=1)
+    gate_list = decompose_one_qubit_gate(Qobj(U), "ZYZ_PauliX")
+
+    # Change the gate targets for ABC and X gates from default = 0
+    # First 2 indices are for A, next 2 are for B and last is for C
+    # Note that we have avoided changing the Pauli X indices because it is
+    # easier to define a gate with the target qubit.
+    target_indices_to_change = [0, 1, 3, 4, 6]
+    for i in target_indices_to_change:
+        gate_list[i].targets = [1]
+
+    # last is phase whose target is not changed based on Lemma 5.2 of
+    # arXiv:quant-ph/9503016v1 but the gate type is changed
+    phase_angle = gate_list[-1].arg_value
+    phase_gate = Gate(
+        "PHASEGATE",
+        targets=[0],
+        arg_value=phase_angle,
+        arg_label=r"{:0.2f} \times \pi".format(phase_angle / np.pi),
+    )
+    Pauli_X = Gate("X", targets=[1], classical_controls=[0])
+
+    CNOT_ctrl_0 = Gate("CNOT", controls=0, targets=1)
+    return (
+        gate_list[0],
+        gate_list[1],
+        Pauli_X,
+        CNOT_ctrl_0,
+        Pauli_X,
+        gate_list[3],
+        gate_list[4],
+        Pauli_X,
+        CNOT_ctrl_0,
+        Pauli_X,  # to do : add a condition to check for control value before
+        # adding Pauli_X
+        gate_list[6],
+        phase_gate,
+    )
+
+
+def _control_0_to_pauli_x(input_gate, num_qubits):
+    """ If the control value is 0 for qubit_i, this function returns info about
+    the indices where pauli X gate should be added.
+    """
+    two_level_gate_info = _two_level_gate_info(input_gate, num_qubits)
+    pauli_x_info = {}
+    for i in range(len(two_level_gate_info)):
+        control_value_list = two_level_gate_info[i]["control_value ="]
+        for j in range(len(control_value_list)):
+            if control_value_list[j] == 0:
+                pauli_x_partial_info = {}
+                gate_indices = [j-1, j]
+                pauli_x_partial_info[j] = gate_indices
+
+            pauli_x_info[i] = pauli_x_partial_info
+
+    return(pauli_x_info)
+
+
+def _multi_cnot_to_two_diff_cnot(input_array, num_qubits):
+    """ Decomposes a multicontrolled cnot into a circuit described by multiple
+    cnots of smaller controls.
+
+    # based on lemma 7.3 of https://arxiv.org/abs/quant-ph/9503016
+    """
+    # add condition to check if input is target for last qubit and rest are
+    # controls - easier with a subclass instance
+
+    # the gate is not a control on num_qubits-1 qubit
+    assert num_qubits >= 5
+    cnot_gate_info = {}
+    num_qubits_list = list(range(num_qubits))
+
+    m = list(range(2, num_qubits-3))
+
+    # m target gates
+
+    # num_qubits-m-1 target gates
+
+
+
+
+
+
+def _decompose_multi_cnot_further():
+    """ Decomposes output of lemma 7.3 even further by using only 1 particular
+    cnot with different controls.
+
+    # based on lemma 7.2 of https://arxiv.org/abs/quant-ph/9503016
+    """
```

Comment:
Some of these functions have incomplete code because locally their output is incorrect. Want to correct the error before committing.